_posts/2016-09-16-angular-2-animation-important-concepts

```diff
@@ -0,0 +1,236 @@
+
+{% highlight js %}
+{% raw %}
+animations: [
+  trigger(’visibilityChanged', [
+  state('shown' , style({ opacity: 1 })), 
+  state('hidden', style({ opacity: 0 }))
+  ])
+]
+{% endraw %}
+{% endhighlight %}
+
+But what does the above mean... and why is this syntax used?
+
+The techno-speak above is saying that when the `visibilityChanged` property changes to the value == 'shown', 
+then the host element opacity changes to 1. And when the value changes == ‘hidden’, the 
+host element opacity should change to 0.
+
+And since we want to animate these changes instead of instantly hiding/showing the content, 
+we now need to configure a *transition*. 
+
+With Angular 2 this is also suprisingly easy:
+
+{% highlight js %}
+{% raw %}
+animations: [
+  trigger(’visibilityChanged', [
+    state('shown' , style({ opacity: 1 })), 
+    state('hidden', style({ opacity: 0 })),
+    transition('* => *', animate('.5s'))
+  ])
+]
+{% endraw %}
+{% endhighlight %}
+
+With the above `transition`, we added information to the animation configuration so each trigger value change will have a 500 msec transition. 
+
+So a **fade-in** (opacity 0 -> 1, with a duration of 500 msec) will occur when the value changes from `hidden` to `shown`. 
+And likewise the **fade-out** (opacity 1 -> 0, with a duration of 500 msec) will occur when the value changes from `shown` to `hidden`.
+
+I this this metadata notation is amazing:
+
+*  intuitive and easy to understand, 
+*  declarative, and 
+*  immediately associated with the component... 
+  *  the `animations` configuration is right above the Class definition!
+
+### Linking Animation to the Component
+
+Now, we are not done yet! While we configured the Animation metadata,  I am sure you are wondering:
+
+*  How is the animation property `visibilityChanged` actually connected to the component ?
+*  How are the animations linked to the component’s properties? 
+
+Since data-binding features are already supported between the **component** and its **template**, 
+Angular 2 uses a <u>special</u> template animation syntax to support triggering the animation after data-binding changes.
+
+So in the component template, we can do this:
+
+{% highlight html %}
+{% raw %}
+<div [@visibilityChanged]="visibility">
+  Can you see me? I should fade in or out...
+</div>
+{% endraw %}
+{% endhighlight %}
+
+
+Above the `@visibilityChanged` is the special template animation property and it uses databinding 
+`[@visibilityChanged]=“visibility”` to bind the component's visibility property to the animation 
+trigger property `visibilityChanged`. 
+
+> This solution of using template binding <u>decouples</u> the animation from the component internals and 
+uses the template as the binding bridge.
+
+And here is the entire component definition updated with Animation features:
+
+{% highlight js %}
+{% raw %}
+import { 
+  Component, OnChanges, Input, 
+  trigger, state, animate, transition, style 
+} from '@angular/core';
```

Comment:
nope. still in `core` because of its very deep integration with the compiler and the code generation processes.